src/kernel/filesystem/ram_device.zig

```diff
@@ -0,0 +1,228 @@
+
+    /// An allocator to allocate memory when needed.
+    allocator: *Allocator,
+
+    /// The memory that the RAM device will control.
+    memory: []u8,
+
+    /// See Device.instance.
+    instance: usize,
+
+    /// See Device.read
+    fn read(device: *const Device, offset: usize, len: usize) Device.Error![]u8 {
+        var self = @fieldParentPtr(RamDevice, "instance", device.instance);
+        if (self.memory.len - len < offset) {
+            return Device.Error.OutOfBounds;
+        }
+        var buff = try self.allocator.alloc(u8, len);
+        std.mem.copy(u8, buff[0..], self.memory[offset .. offset + len]);
+        return buff;
+    }
+
+    /// See Device.write
+    fn write(device: *const Device, offset: usize, bytes: []const u8) Device.Error!void {
+        var self = @fieldParentPtr(RamDevice, "instance", device.instance);
+        if (self.memory.len - bytes.len < offset) {
+            return Device.Error.OutOfBounds;
+        }
+        std.mem.copy(u8, self.memory[offset..], bytes[0..]);
+    }
+
+    ///
+    /// De-initialise a RAM device. This will free the underlying device, the memory this device
+    /// controls and this device itself.
+    ///
+    /// Argument:
+    ///     IN self: *Self - Self.
+    ///
+    pub fn deinit(self: *Self) void {
+        self.allocator.destroy(self.device);
+        self.allocator.free(self.memory);
+        self.allocator.destroy(self);
+    }
+
+    ///
+    /// Initialise a new RAM device with the memory provided duplicated. After this is called, the
+    /// memory provided can be freed.
+    ///
+    /// Arguments:
+    ///     IN memory: []u8          - The pointer to the bytes that the RAM device will manage.
+    ///     IN allocator: *Allocator - The allocator when needing to allocate memory.
+    ///
+    /// Return: *RamDevice
+    ///     A new RAM device.
+    ///
+    /// Error: Allocator.Error
+    ///     error.OutOfMemory - If there isn't enough memory when initialising the RAM device.
+    ///                         Any memory allocated will be freed on return.
+    ///
+    pub fn init(memory: []u8, allocator: *Allocator) Allocator.Error!*RamDevice {
+        const ram_device = try allocator.create(RamDevice);
+        errdefer allocator.destroy(ram_device);
+        const device = try allocator.create(Device);
+        errdefer allocator.destroy(device);
+
+        ram_device.* = .{
+            .device = device,
+            .allocator = allocator,
+            // QUESTION: Should we dupe the memory? We could just have the memory with the same life time as the pointer passed (caller)
+            //           Then just ass a comment in the doc comment about the life time.
+            //           This would improve speed and memory usage, but the device should control the memory.
+            //           Could add 2 init function, one with a size to allocate and one with a memory pointer (this one).
+            .memory = try allocator.dupe(u8, memory),
```

Comment:
My understanding is that this RAM device receives a region of memory to manage, rather than creating its own region? I think taking control of the given slice would make this more flexible and would cut down on allocations.

Reply:
Yeh, but my concern was that if the caller frees the memory and still tries to use the device will cause undefined behavior. So will have a comment saying not to free the memory.